articles/events_executor.md

```diff
@@ -0,0 +1,233 @@
+ - Inefficiency of the timers management.
+   This is currently done in the rcl layer, where at each iteration the full list of timers associated with an executor is checked twice.
+ - The presence of several abstraction layers that do not simply forward data, but rather perform non trivial operations, between the application and the underlying middleware.
+ 
+The CPU overhead of ROS 2 with respect to other pub/sub systems (as the same DDS middlewares that it uses under the hood), makes it difficult to integrate ROS 2 on embedded platforms with constrained resources, as a not negligible amount of CPU is needed just for the executor to identify entities with work to do.
+ 
+## Requirements
+ 
+The following requirements have been taken account for the design of a new ROS 2 executor:
+ - Good performance: the executor shall not be the CPU bottleneck of a ROS 2 system, thus allowing to obtain performance comparable to the ones of the underlying middleware that is being used.
+ - Scalability: The executor must perform well independently of the system size, e.g. the latency for processing a fixed amount of messages must be independent from the number of entities.
+   This can flow into the "You don't pay for what you don't use" paradigm.
+ - Extensibility: The executor should set a base from which it is straight-forward to add more functionalities without impacting performances or requiring heavy modifications of the ROS2 core libraries.
+ - Ordering of events: The executor should process events in the same order they have occurred.
+ - High precision and accurate timers: The executor should provide mechanisms to ensure that timers are executed with the minimum possible latency.
+ 
+## The Events Executor
+
+In order to address the aforementioned requirements, we designed the so-called `EventsExecutor`.
+The `EventsExecutor` inherits from the base `Executor` class.
+The main difference between this and the other executor implementations is that the `EventsExecutor` does not use the concept of a waitset.
+The `EventsExecutor` is characterized by two main components, i.e. the `EventsQueue` and the `TimersManager`.
+
+### Brief Overview
+
+In order to execute generic ROS 2 entities (i.e. everything except timers), the `EventsExecutor` relies on an event queue.
+Whenever a ROS 2 entity that is associated to the executor has work to do, it will push an event into the executor's queue.
+Then the executor can process these events in a FIFO manner, without need for expensive entities look-ups.
+Processing an event results in different operations depending on the entity that generated it.
+
+Timers are executed in a separate task by a timers manager, where they are kept in a priority queue sorted according to their expiration time. The task then has to monitor only the first element of the queue and can execute its callback as soon as it expires.
```

Comment:
Yes, the TimersManager is on purpose a separate entity from the executor to have multiple parts of the system (or even application themselves) to take advantage of it.

Reply:
Though it reaches beyond the scope of this PR, bear in mind that eventually the `TimersManager` class will have to deal with different timers associated with different clocks. Perhaps a queue per clock type? Synchronous API for `spin_once()` implementation will probably suffer though.

Reply:
Could you clarify what's the use-case for having timers with different clock coexist in the same executor?

Reply:
> Could you clarify what's the use-case for having timers with different clock coexist in the same executor?

The simplest one I can think of is a node that unconditionally requires a timer to be associated with the system clock (e.g. to implement some I/O timeout) but otherwise associates all other timers to the ROS clock for proper operation in simulation whenever RTF != 1.

Reply:
How do the existing executors handle this situation?
In general I think that mixing timers with different clocks in the same execution context can easily lead to errors.

Reply:
They don't :). Thus the disclaimer above:

> Though it reaches beyond the scope of this PR, bear in mind that eventually the TimersManager class will have to deal with different timers...

It's a known limitation that hopefully we can overcome at some point in the future.